libraries/AP_Frsky_Telem/AP_Frsky_SPort.cpp

```diff
@@ -0,0 +1,253 @@
+                    break;
+                case 6:
+                    send_sport_frame(SPORT_DATA_FRAME, DATA_ID_GPS_COURS_BP, _SPort_data.yaw); // send heading in degree based on AHRS and not GPS
+                    _SPort.gps_refresh = true;
+                    break;
+                }
+                if (++_SPort.gps_call > 6) {
+                    _SPort.gps_call = 0;
+                }
+                break;
+            case SENSOR_ID_SP2UR: // Sensor ID  6
+                switch (_SPort.various_call) {
+                case 0 :
+                    send_sport_frame(SPORT_DATA_FRAME, DATA_ID_TEMP2, (uint16_t)(AP::gps().num_sats() * 10 + AP::gps().status())); // send GPS status and number of satellites as num_sats*10 + status (to fit into a uint8_t)
+                    break;
+                case 1:
+                    send_sport_frame(SPORT_DATA_FRAME, DATA_ID_TEMP1, gcs().custom_mode()); // send flight mode
+                    break;
+                }
+                if (++_SPort.various_call > 1) {
+                    _SPort.various_call = 0;
+                }
+                break;
+            }
+            _SPort.sport_status = false;
+        }
+    }
+}
+
+/*
+ * send an 8 bytes SPort frame of FrSky data - for FrSky SPort protocol (X-receivers)
+ */
+void  AP_Frsky_SPort::send_sport_frame(uint8_t frame, uint16_t appid, uint32_t data)
+{
+    uint8_t buf[8];
+
+    buf[0] = frame;
+    buf[1] = appid & 0xFF;
+    buf[2] = appid >> 8;
+    memcpy(&buf[3], &data, 4);
+
+    uint16_t sum = 0;
+    for (uint8_t i=0; i<sizeof(buf)-1; i++) {
+        sum += buf[i];
+        sum += sum >> 8;
+        sum &= 0xFF;
+    }
+    sum = 0xff - ((sum & 0xff) + (sum >> 8));
+    buf[7] = (uint8_t)sum;
+
+    // perform byte stuffing per SPort spec
+    uint8_t len = 0;
+    uint8_t buf2[sizeof(buf)*2+1];
+
+    for (uint8_t i=0; i<sizeof(buf); i++) {
+        uint8_t c = buf[i];
+        if (c == FRAME_DLE || buf[i] == FRAME_HEAD) {
+            buf2[len++] = FRAME_DLE;
+            buf2[len++] = c ^ FRAME_XOR;
+        } else {
+            buf2[len++] = c;
+        }
+    }
+#ifndef HAL_BOARD_SITL
+    /*
+      check that we haven't been too slow in responding to the new
+      UART data. If we respond too late then we will overwrite the next
+      polling frame.
+      SPort poll-to-pool period is 11.65ms, a frame takes 1.38ms
+      this leaves us with up to 10ms to respond but to play it safe we
+      allow no more than 7500us
+     */
+    uint64_t tend = _port->receive_time_constraint_us(1);
+    uint64_t now = AP_HAL::micros64();
+    uint64_t tdelay = now - tend;
+    if (tdelay > 7500) {
+        // we've been too slow in responding
+        return;
+    }
+#endif
+    _port->write(buf2, len);
+}
+
+/*
+ * prepare value for transmission through FrSky link
+ * for FrSky SPort Passthrough (OpenTX) protocol (X-receivers)
+ */
+uint16_t AP_Frsky_SPort::prep_number(int32_t number, uint8_t digits, uint8_t power)
```

Comment:
this is only used by passthrough so should probably go in the relevant backend.
As a side note the PR for scripting support will need public access to this method

Reply:
Fixed.